torch/_weights_only_unpickler.py

```diff
@@ -122,23 +138,85 @@ def _get_allowed_globals():
     return rc
 
 
+class _Unframer:
+
+    def __init__(self, file_read, file_readline, file_tell=None):
+        self.file_read = file_read
+        self.file_readline = file_readline
+        self.current_frame = None
+
+    def readinto(self, buf):
+        if self.current_frame:
+            n = self.current_frame.readinto(buf)
+            if n == 0 and len(buf) != 0:
+                self.current_frame = None
+                n = len(buf)
+                buf[:] = self.file_read(n)
+                return n
+            if n < len(buf):
+                raise UnpicklingError(
+                    "pickle exhausted before end of frame")
+            return n
+        else:
+            n = len(buf)
+            buf[:] = self.file_read(n)
+            return n
```

Comment:
Nits (type annotations, avoid code duplication)
```suggestion
    def readinto(self, buf) -> int:
        def readinto_noframe() -> int:
                n = len(buf)
                buf[:] = self.file_read(n)
                return n
        if not self.current_frame:
            return readinto_noframe()
        n = self.current_frame.readinto(buf)
        if n == 0 and len(buf) != 0:
            self.current_frame = None
            return readint_noframe()
        if n < len(buf):
            raise UnpicklingError("pickle exhausted before end of frame")
        return n
```

Reply:
Done